### ISSUE TITLE
Enhancement : add automatic allocation of categories

### ISSUE THREAD

**Issue by ZeHiro** [bnjbvr/kresus#41]
Hello,

It could be nice to have automatic allocation of categories base on : 
- regex on operation name
- type of transation (cach wthdraw, card payment, check etc)

Thanks

**Comment by jankeromnes** [comment]
Great idea! I think automatic categorization is something @bnjbvr eventually planned to do. He was even talking about machine learning at some point.

**Comment by bnjbvr** [comment]
Indeed! First automatic allocation will work with regexp and manual creation of rules. Ideally, rules detection could be made automatic easily...

**Comment by hamzux** [comment]
hello, 
Is having a predefined list of shops (classified by category) an option to start implementing this feature?

For example: 

Check all expenses containing E. Leclrec, Carrefour, Auchamp etc as Groceries (even if it's not always what you buy at these venues).
We would end up with a reference database for each country. 

Has someone experimented with machine learning?

Thanks.

**Comment by bnjbvr** [comment]
Hi @hamzux,
That could work as a start, but there is one tiny problem: we don't have default categories, so we can't have a static mapping of regexp to categories. Categories are variables. What this means is that we need to start with a new data type, containing the following information: Premise => Consequence, enabled or not.

`Premise` would be an object containing one of the fields `labelContains`, or `typeIs`, or `dateIs`, or `dateAround`. `Consequence` would be an object containing all the values that are automatically assigned to that operation, if the premise is verified: e.g. it would contain `category`, or `type`, etc.
The `enabled` field would be a three state boolean (disabled, always enabled, enabled with confirm). We can start with a simple boolean (by not having "enabled with confirm" for a start).

Let's say I have the rule:

```
{premise:
  {
    labelContains: "Leclerc"
  },
  consequence:
  {
    category: 38f39ab9c9d7361a9c7d6728 // that's a category id
  },
  enabled: true
}
```

Then an operation which label is `COURSES LECLERC 21/01 CB 1234` would get automatically assigned the category with the specified id.

That creates an implicit dependency between categories/types and rules, so when a category is deleted, we have to ensure the rule is deleted as well (or even better, we just ignore the consequence in this case).

How does that sound?

**Comment by bnjbvr** [comment]
Forgot to say: bayesian filters could be used to infer rules from the manual classification, but would be too imprecise otherwise (if enabled without asking for confirmation).

**Comment by hamzux** [comment]
Hello,

I managed to setup kresus correctly in standalone mode. I've created a new method called classify in operations controller which loops through all operations in all accounts and tests if operation.title = 'E.LECLERC'. Also added route 'operations/classify' { get: operations.classify}.
When I call localhost:9876/operations/classify I get a 404 error.

Did I miss something ?
Thanks

**Comment by bnjbvr** [comment]
Hi, it's a bit hard to tell without looking at the code. Can you push your WIP on a branch of your fork, so that I can give it a look, please?
Cheers

**Comment by SomeAverageDev** [comment]
hello @hamzux,
could you share your code ? we might have a look to move forward on this great feature !
